### PR TITLE
feat: support load and query with bm25 metric

### PR DESCRIPTION
**Pull request by aoiasd**
relate: https://github.com/milvus-io/milvus/issues/35853

### CODE REVIEW

File: internal/querynodev2/delegator/idf_oracle.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+		o.current.Merge(stats)
+	case segments.SegmentTypeSealed:
+		if _, ok := o.sealed[segmentID]; ok {
+			return
+		}
+		o.sealed[segmentID] = &bm25Stats{
+			stats:         stats,
+			activate:      false,
+			targetVersion: initialTargetVersion,
+		}
```
Comment (zhengbuqian) [inline internal/querynodev2/delegator/idf_oracle.go:128]: should we call `o.current.Merge(stats)`?

File: internal/querynodev2/pipeline/embedding_node.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+func GetEmbeddingFieldData(datas []*schemapb.FieldData, fieldID int64) ([]string, error) {
+	for _, data := range datas {
+		if data.GetFieldId() == fieldID {
+			return data.GetScalars().GetStringData().GetData(), nil
+		}
+	}
+	return nil, fmt.Errorf("field%d not found", fieldID)
+}
+
+func GetSparseVectorDim(data [][]byte) int64 {
```
Comment (zhengbuqian) [inline internal/querynodev2/pipeline/embedding_node.go:231]: function is unused. If necessary, place it in file `SparseFloatRowDim` is defined

**Review by zhengbuqian — COMMENTED** [review]
thanks!!!

File: internal/querynodev2/delegator/distribution.go [hunk 3]
```diff
@@ -74,6 +74,8 @@ type distribution struct {
 	// current is the snapshot for quick usage for search/query
 	// generated for each change of distribution
 	current *atomic.Pointer[snapshot]
+
+	idfOracle IDFOracle
```
[the remark this answers is not in the document]
Comment (aoiasd) [reply internal/querynodev2/delegator/distribution.go:78]: IDFOracle was Interface

File: internal/querynodev2/delegator/idf_oracle.go (hunk 1 above)
Comment (aoiasd) [reply internal/querynodev2/delegator/idf_oracle.go:136]: No, sealed segment should be deactivated till sync distribution.

File: internal/querynodev2/segments/collection.go [hunk 4]
```diff
@@ -299,6 +299,18 @@ func NewCollectionWithoutSchema(collectionID int64, loadType querypb.LoadType) *
 	}
 }
 
+// new collection without segcore prepare
+// ONLY FOR TEST
+func NewCollectionWithoutSegcore(collectionID int64, schema *schemapb.CollectionSchema) *Collection {
```
Comment (zhengbuqian) [inline internal/querynodev2/segments/collection.go:304]: reminder: do we need to delete this

**Review by zhengbuqian — COMMENTED** [review]
thanks!

mostly looks good, but need to handle hybrid search.

File: internal/querynodev2/segments/collection.go (hunk 4 above)
Comment (aoiasd) [reply internal/querynodev2/segments/collection.go:304]: All Collection struct field was private, so we must provide a special create function if some test in other package need use meta without segcore.

File: internal/proxy/task_index.go [hunk 5]
```diff
@@ -361,8 +361,12 @@ func (cit *createIndexTask) parseIndexParams() error {
 				return merr.WrapErrParameterInvalid("valid index params", "invalid index params", "float vector index does not support metric type: "+metricType)
 			}
 		} else if typeutil.IsSparseFloatVectorType(cit.fieldSchema.DataType) {
-			if metricType != metric.IP {
-				return merr.WrapErrParameterInvalid("valid index params", "invalid index params", "only IP is the supported metric type for sparse index")
+			if metricType != metric.IP && metricType != metric.BM25 {
+				return merr.WrapErrParameterInvalid("valid index params", "invalid index params", "only IP&BM25 is the supported metric type for sparse index")
+			}
+
+			if metricType == metric.BM25 && cit.functionSchema.GetType() != schemapb.FunctionType_BM25 {
```
[the remark this answers is not in the document]
Comment (aoiasd) [reply internal/proxy/task_index.go:368]: GetType will return nil if functionSchema was nil, so we can skip check if it was exist

File: internal/proxy/task_search.go [hunk 6]
```diff
@@ -490,38 +492,38 @@ func (t *searchTask) initSearchRequest(ctx context.Context) error {
 	return nil
 }
 
-func (t *searchTask) tryGeneratePlan(params []*commonpb.KeyValuePair, dsl string) (*planpb.PlanNode, *planpb.QueryInfo, int64, error) {
+func (t *searchTask) tryGeneratePlan(params []*commonpb.KeyValuePair, dsl string) (*planpb.PlanNode, *planpb.QueryInfo, *schemapb.FieldSchema, int64, error) {
```
Comment (czs007) [inline internal/proxy/task_search.go:495]: how about adding fieldID to planpb.QueryInfo?

File: internal/querynodev2/delegator/idf_oracle.go [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+func (s *bm25Stats) GetStats(fieldID int64) (*storage.BM25Stats, error) {
+	stats, ok := s.stats[fieldID]
+	if !ok {
+		return nil, fmt.Errorf("field not found in idf oracle BM25 stats")
+	}
+	return stats, nil
+}
+
+func (s *bm25Stats) NumRow() int64 {
```
Comment (czs007) [inline internal/querynodev2/delegator/idf_oracle.go:80]: add parameter fieldID ?

File: internal/proxy/task_search.go (hunk 6 above)
Comment (aoiasd) [reply internal/proxy/task_search.go:495]: Seems we don't transfer QueryInfo to segcore or delegator.

File: internal/querynodev2/delegator/idf_oracle.go (hunk 7 above)
Comment (aoiasd) [reply internal/querynodev2/delegator/idf_oracle.go:80]: Seems not necessary.

File: internal/querynodev2/delegator/delegator_data.go [hunk 8]
```diff
@@ -963,3 +993,42 @@ func (sd *shardDelegator) TryCleanExcludedSegments(ts uint64) {
 		sd.excludedSegments.CleanInvalid(ts)
 	}
 }
+
+func (sd *shardDelegator) buildBM25IDF(req *internalpb.SearchRequest) (float64, error) {
+	pb := &commonpb.PlaceholderGroup{}
+	proto.Unmarshal(req.GetPlaceholderGroup(), pb)
+	holder := pb.Placeholders[0]
```
Comment (czs007) [inline internal/querynodev2/delegator/delegator_data.go:1005]: remember to check whether pb.Placeholders is empty.

File: internal/querynodev2/pipeline/embedding_node.go [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+		if err != nil {
+			return nil, err
+		}
+		node.functionRunners = append(node.functionRunners, functionRunner)
+	}
+	return node, nil
+}
+
+func (eNode *embeddingNode) Name() string {
+	return fmt.Sprintf("embeddingNode-%s-%s", "BM25test", eNode.channel)
```
Comment (czs007) [inline internal/querynodev2/pipeline/embedding_node.go:78]: Rename "BM25test" to "BM25"?

File: internal/querynodev2/pipeline/embedding_node.go [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+		log.Error(err.Error(), zap.Int64("collectionID", eNode.collectionID), zap.String("channel", eNode.channel))
+		return err
+	}
+
+	if iData.InsertRecord == nil {
+		iData.InsertRecord = insertRecord
+	} else {
+		err := typeutil.MergeFieldData(iData.InsertRecord.FieldsData, insertRecord.FieldsData)
+		if err != nil {
+			log.Error("failed to merge field data", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (czs007) [reply internal/querynodev2/pipeline/embedding_node.go:113]: Level can be WARN

[the remark this answers is not in the document]
Comment (czs007) [reply internal/querynodev2/pipeline/embedding_node.go:113]: log.Error means fatal error which the program can not run normally.
log.Warn indicates a need for maintenance personnel to be vigilant or attentive.

File: internal/querynodev2/segments/collection.go (hunk 4 above)
Comment (czs007) [inline internal/querynodev2/segments/collection.go:304]: Or, you can rename it to NewCollectionWithoutSegcoreForTest

**Pull request merged**